### ISSUE TITLE
Incompatible with AngularJS

### ISSUE THREAD

**Issue by DavidPrevot** [tchwork/jsqueeze#18]
Hi,

Using JSqueeze on ownCloud, it fails to minify correctly the [JavaScript file](http://tilapin.org/tmp/assets-full.js) provided by the [Notes application](https://github.com/owncloud/notes) with the [following error](https://docs.angularjs.org/error/$injector/modulerr?p0=Notes&p1=[$injector:unpr]%20Unknown%20provider:%20t%0Ahttp:%2F%2Ferrors.angularjs.org%2F1.4.0-build.3841%2Bsha.ef894c8%2F$injector%2Funpr%3Fp0%3Dt%0Ap%2F%3C@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:418%0Abe%2Fp.$injector%3C@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:24811%0Ai@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:26521%0Ar@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:26869%0Ao@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:25886%0Abe%2Fb%2F%3C@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:25996%0Ar@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:1399%0Ab@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:25764%0Abe@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:24955%0AVt%2Fu@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:8291%0AVt@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:8607%0Aso@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:7842%0A@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:154164%0AjQuery.Callbacks%2Ffire@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:2:9819%0AjQuery.Callbacks%2Fself.fireWith@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:2:11122%0A.ready@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:1:5073%0Acompleted@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:1:1214%0A):

    Error: [$injector:modulerr] Failed to instantiate module Notes due to:
    [$injector:unpr] Unknown provider: t
    http://errors.angularjs.org/1.4.0-build.3841+sha.ef894c8/$injector/unpr?p0=t
    p/<@http://localhost/owncloud/assets/72773572308bad612ffce802e84db88d.js:104:418
    be/p.$injector<@http://localhost/owncloud/assets/72773572308bad612ffce802e84db88d.js:104:24811
    [ … more lines … ]
    ...="+encodeURIComponent(qt(n[r]))};return new t(i)}};var Xi=/^\/(.+)\/([a-z]*)$/,e...
    -------------------------------------------^

(Firebug points the 418th character of the 104th line)

The unminified counterparts it comes from is:

    for (i = SKIP_INDEXES, paramPrefix = '?'; i < templateArgs.length; i++, paramPrefix = '&') {
      message += paramPrefix + 'p' + (i - SKIP_INDEXES) + '=' +
        encodeURIComponent(toDebugString(templateArgs[i]));
    }

    return new ErrorConstructor(message);

I don’t believe the actual failure comes from here anyway: it’s just the error handling mechanism, as are the first hops (104:24811, 104:26521, etc.) pointed from the full [debug message](https://docs.angularjs.org/error/$injector/modulerr?p0=Notes&p1=[$injector:unpr]%20Unknown%20provider:%20t%0Ahttp:%2F%2Ferrors.angularjs.org%2F1.4.0-build.3841%2Bsha.ef894c8%2F$injector%2Funpr%3Fp0%3Dt%0Ap%2F%3C@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:418%0Abe%2Fp.$injector%3C@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:24811%0Ai@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:26521%0Ar@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:26869%0Ao@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:25886%0Abe%2Fb%2F%3C@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:25996%0Ar@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:1399%0Ab@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:25764%0Abe@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:24955%0AVt%2Fu@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:8291%0AVt@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:8607%0Aso@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:7842%0A@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:104:154164%0AjQuery.Callbacks%2Ffire@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:2:9819%0AjQuery.Callbacks%2Fself.fireWith@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:2:11122%0A.ready@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:1:5073%0Acompleted@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:1:1214%0A).

It may be related to the loading of angular.module (I really don’t understand much about JavaScript, sorry) as initiated by [config/app.js](https://github.com/owncloud/notes/blob/master/js/config/app.js) and the way JSqueze shrinks this call and the actual AngularJS handling.

The three files corresponding to the [three minified lines (104 to 106)](http://tilapin.org/tmp/assets.js) are:
- [angular.js](https://github.com/owncloud/notes/blob/master/js/vendor/angular/angular.js)
- [angular-route.js](https://github.com/owncloud/notes/blob/master/js/vendor/angular-route/angular-route.js)
- [app.js](https://github.com/owncloud/notes/blob/master/js/public/app.js)

(Replacing those three lines by the content of those three files makes the application usable again.)

Please note that I’ve also reproduced the issue with two other versions (1.2.28 and 1.3.13) of AngularJS. I’d be happy if you could help finding a workaround if the issue is tough to fix. I’d also be happy to provide a more actionable bug report if you point me in the right direction (I have no idea how to properly debug such issue right now).

**Comment by nicolas-grekas** [comment]
can you disable $singleLine mode? That make debugging easier...

**Comment by DavidPrevot** [comment]
> Can you please disable $singleLine mode?

Sure, thanks for the tip. Please follow the links to the updated [debug message](https://docs.angularjs.org/error/$injector/modulerr?p0=Notes&p1=[$injector:unpr]%20Unknown%20provider:%20t%0Ahttp:%2F%2Ferrors.angularjs.org%2F1.4.0-build.3841%2Bsha.ef894c8%2F$injector%2Funpr%3Fp0%3Dt%0Ap%2F%3C@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13463:8%0Abe%2Fp.$injector%3C@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13928:7%0Ai@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13965:13%0Ar@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13972:36%0Ao@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13954:1%0Abe%2Fb%2F%3C@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13958:1%0Ar@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13475:43%0Ab@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13950:1%0Abe@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13930:3%0AVt%2Fu@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13630:7%0AVt@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13636:1%0Aso@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:13619:1%0A@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:16880:23%0AjQuery.Callbacks%2Ffire@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:2:9819%0AjQuery.Callbacks%2Fself.fireWith@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:2:11122%0A.ready@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:1:5073%0Acompleted@http:%2F%2Flocalhost%2Fowncloud%2Fassets%2F72773572308bad612ffce802e84db88d.js:1:1214%0A) and full [JavaScript file](http://tilapin.org/tmp/assets-full-nosl.js).

**Comment by nicolas-grekas** [comment]
I see no trivial pb for now, I'll need to dig deeper.
Any way I can run something an reproduce the issue locally?
Did you verify it works without JSqueeze? (by e.g. making JSqueeze return the original input string?)

**Comment by DavidPrevot** [comment]
> Any way I can run something an[d] reproduce the issue locally?

I “just” installed the [latest ownCloud](https://owncloud.org/#getapps) (8.0.2), [patched it](https://github.com/AdamWill/core/commit/da866a885bbc035a12c4e98fd0eed2ec660b0d7f) to use JSqueeze instead of JSMin, and installed the [Notes application](https://github.com/owncloud/notes) that is unusable because of this issue (I’d expect any application using AngularJS —like [News](https://apps.owncloud.com/content/show.php/News?content=168040), [Music](https://apps.owncloud.com/content/show.php/Music?content=164319) or [Tasks](https://apps.owncloud.com/content/show.php/Tasks?content=164356)– to also fail the same way).

If you’re running Debian (or a derivative), I’m happy to provide the (currently WIP) packages I’m preparing, where I’m getting this issue. If you’re running Fedora (or a derivative), I guess @AdamWill may also help providing WIP packages.

I can also set up such a (broken) instance and give you a web access to it if that can help (I wish I could provide a minimalistic way to reproduce it, but I’m currently lost in JavaScript, maybe this weekend if I manage to learn how to use AngularJS…)

> Did you verify it works without JSqueeze?

Yes, that’s what I meant by “Replacing those three lines by the content of those three files makes the application usable again.”

**Comment by AdamWill** [comment]
Yep, I can indeed provide Fedora or EL 7 packages with JSqueeze enabled if that'd be useful.

**Comment by nicolas-grekas** [comment]
I'm used to Debian, so I could try with your packages on a VM

**Comment by DavidPrevot** [comment]
The packages not yet in the official Debian archive are on https://people.debian.org/~taffit/owncloud/.

You may add the following line to `/etc/apt/sources.list` (as well as `experimental`), and simply install `owncloud-notes` to get everything needed:

    deb https://people.debian.org/~taffit/owncloud ./

(The `apt-transport-https` package is needed to use HTTPS repositories.)

Once installed, ownCloud should be available on `http://localhost/owncloud`. You need to activate the Notes application, and then, in the Notes interface, you should reproduce the issue.

Feel free to ping me or IRC if it you have any trouble setting up this test environment (I’m `taffit` in at least OFTC and FreeNode).

**Comment by DavidPrevot** [comment]
Here is a simple way to reproduce an error with AngularJS: with the simplified sample `app/index.html` from the beginning of their [tutorial](https://docs.angularjs.org/tutorial/step_00):

```html
<!doctype html>
<html lang="en" ng-app>
<head>
  <script src="angular.js"></script>
</head>
<body>
{{'Yay' + '!'}}
</body>
</html>
```

Using an unminified version of `angular.js`, it correctly displays ` Yay! `, but with a minified by JSqueeze version of `angular.js`, it displays ` {{'Yay' + '!'}} ` with an error in the FireBug JS console.

Hope this helps: even if the error message is different in the console, it shows at least an incompatibility, and involves a lot less to reproduce (even if `angular.js` is big…)

(Also, a minified by uglifyjs version of `angular.js` works fine, but I noticed it’s a lot bigger: original is 968K, JSqueeze’s version is 152K, while uglifyjs’ version is 263K).

**Comment by nicolas-grekas** [comment]
Thanks! Comparing uglifyjs' vs jsqueeze's will help for sure!

**Comment by nicolas-grekas** [comment]
I managed to create a currently failing test in #19
Fix pending :)

**Comment by nicolas-grekas** [comment]
Btw, I get 145k with Uglify on my side

**Comment by DavidPrevot** [comment]
> I managed to create a currently failing test in #19

Yay ;)
> Btw, I get 145k with Uglify on my side

Maybe the version currently in Debian is not as effective as the latest upstream one…

**Comment by nicolas-grekas** [comment]
Fixed in #19

**Comment by DavidPrevot** [comment]
[ Do not hesitate (to tell me) to open another issue if you prefer. ]

##### `$scope` should be protected

Still from their [tutorial](https://docs.angularjs.org/tutorial/step_02): [controler.js](https://github.com/angular/angular-phonecat/blob/step-2/app/js/controllers.js) is broken if handled via JSqueeze:

``` JavaScript
[…]
phonecatApp.controller('PhoneListCtrl', function($scope) {
  $scope.phones = [
[…]
```
becomes
``` JavaScript
phonecatApp.controller('PhoneListCtrl',function(t){t.phones=
```
then “Error: [$injector:unpr] Unknown provider: tProvider <- t <- PhoneListCtrl” is thrown in the JavaScript console and the page is not displayed correctly (UglifyJS doesn’t change `$scope` in `t`). That’s not exactly the same error as the one pointed in the beginning, but it’s getting closer.

**Comment by DavidPrevot** [comment]
##### Possible workaround

Hi,

I may have found a way to workaround this issue (or maybe that was just me misusing JavaScript…): I was actually using a [plain concatenation](https://github.com/owncloud/notes/blob/master/js/public/public/app.js#L104) instead of the [improved (or protected) one](https://github.com/owncloud/notes/blob/master/js/public/app.js#L103). The differences are of this kind:

```diff
-app.controller('AppController', function ($scope, $location, is) {
+app.controller('AppController', ["$scope", "$location", "is", function ($scope, $location, is) {
```

Using the improved (or protected) version seems to work fine with JSqueeze, so I’m back being a happy user and tester (I suppose the issue is still worth fixing if it’s really an issue rather than a misuse on my side – in that case I apologize for being a bad tester and user).

**Comment by nicolas-grekas** [comment]
This is it! Angular even has a doc about this new syntax being friendly to
minifiers. Not sure why Uglify works, maybe it has some rule for angular.
But for sure this is required for JSqueeze! Congrats, one bug fixed on each
side :-)

**Comment by nicolas-grekas** [comment]
Here is the doc pointer: https://docs.angularjs.org/tutorial/step_05#a-note-on-minification

**Comment by nicolas-grekas** [comment]
Fixed by #19

**Comment by nicolas-grekas** [comment]
v2.0.2 and v1.0.7 tagged with fix embedded
